types/react-dom/canary.d.ts

```diff
@@ -32,21 +32,64 @@ import ReactDOM = require(".");
 export {};
 
 declare module "." {
-    type PreloadAs = "font" | "script" | "style";
+    function prefetchDNS(href: string): void;
+
+    interface PreconnectOptions {
+        // Don't create a helper type.
+        // It would have to be in module scope to be inlined in TS tooltips.
+        // But then it becomes part of the public API.
+        // TODO: Upstream to microsoft/TypeScript-DOM-lib-generator -> w3c/webref
+        // since the spec has a notion of a dedicated type: https://html.spec.whatwg.org/multipage/urls-and-fetching.html#cors-settings-attribute
+        crossOrigin?: "anonymous" | "use-credentials" | "" | undefined;
+    }
+    function preconnect(href: string, options?: PreconnectOptions): void;
+
+    type PreloadAs = "font" | "image" | "script" | "style";
```

Comment:
Internally we just treat this as a string argument now however what probably make sense in TS is to enumerate all of the known valid as types for preloads

```
// from: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/preload#what_types_of_content_can_be_preloaded

audio
document
embed
fetch
object
script
style
track
worker
video
```